lib,zebra,sharpd: allow recursive resolution of nexthop-groups

mjstapp opened the pull request:
Add initial support for recursive resolution of nexthop-groups:

- add a flags field to the nexthop-group zapi message, and assign a RECURSION flag value
- add a "recursive" config option to the nexthop-group config submode
- use the config and zapi changes in sharpd
- in zebra, add code to iterate over a group's nexthops, use the zebra validity-checking logic, and update as necessary. This is driven by daemon-owned nhgs, allowing the use of recursive nexthops in those nhgs.

I'm making this a draft for now, so we can have some discussion. There are still some gaps - sharpd's groups aren't being uninstalled and deleted properly, for example, so I'd like to clean that up. I'm not sure that all necessary changes to a nexthop are being detected in this path. And I haven't exercised these nexthop-group changes in coordination with routes, so there are some questions and TODOs there also.

Commit:
zebra: reorg api used to add daemon-owned nhg

Re-org/simplify the api used to add nhgs via zapi messages
from daemons, using a temporary nhe so we can convey all nhe
attributes.

Signed-off-by: Mark Stapp <mjs@voltanet.io>

Commit:
lib,zebra: add flags to NHG zapi

Add a flags field to the zapi nexthop-group struct; encode
and decode the field.

Signed-off-by: Mark Stapp <mjs@cisco.com>

Commit:
lib: add recursive config option for nexthop-groups

Add a 'recursive' flag field and corresponding config option
for nexthop-groups.

Signed-off-by: Mark Stapp <mjs@cisco.com>

Commit:
sharpd: revise nexthop-group api to carry config info

Update sharpd's internal api so that we carry more config info
when processing nhg config callbacks. This is used to allow
access to, e.g., the flags field.

Signed-off-by: Mark Stapp <mjs@cisco.com>

Commit:
zebra: perform recursive resolution for nexthop-groups

Add support for recursive resolution of nexthop-groups: iterate
over a group's nexthops, use the zebra validity-checking logic,
and update as necessary. This is driven by daemon-owned nhgs,
allowing the use of recursive nexthops in those nhgs.
Daemon NHGs that don't request resolution use the pre-existing
behavior (zebra just tries to install what the daemon sends.)

Signed-off-by: Mark Stapp <mjs@cisco.com>

ton31337 commented on lib/nexthop_group.c:
```diff
@@ -706,6 +706,30 @@ DEFPY(no_nexthop_group_resilience,
 	return CMD_SUCCESS;
 }
 
+DEFPY(nexthop_group_recursive,
```
Can we document this?

mjstapp replied:
doh! yes, of course
this is really just a draft, so I know there are still some rough edges...

mjstapp replied:
pushed it...

Commit:
doc: add doc for nexthop-group recursive config

Add some words about the recursive nhg config option; clean
up duplicate doc tags.

Signed-off-by: Mark Stapp <mjs@cisco.com>

pguibert6WIND commented on lib/nexthop_group.c:
```diff
@@ -706,6 +706,30 @@ DEFPY(no_nexthop_group_resilience,
+
+	flags = nhgc->flags;
+
+	if (no)
```
is it possible to add a topotest like was proposed in https://github.com/FRRouting/frr/pull/16028/commits/e60cd344d5c26c26b0e6bcaac169a4420bbd21f4 ?

pguibert6WIND reviewed (commented):
it is possible to see how to reuse some commits from https://github.com/FRRouting/frr/pull/16028/ ?
As discussed yesterday, this extension has not wanted to control the zebra NHGs. The recursive keyword has been introduced, along with testing and many other zebra elements.

pguibert6WIND commented:
as said on slack, a similar proposal is done at https://github.com/FRRouting/frr/pull/16947.
please review.

pguibert6WIND commented:
> * please no AFI changes in this pull request. lets focus on recursive only.
> 
>     * missing topotests
 
no feedback ?

mjstapp commented:
rebased to resolve merge conflicts
